robots/bebop.h

```diff
@@ -0,0 +1,1188 @@
+                 * Notify waiting threads. The if statement is necessary so that
+                 * we don't report the current maximum before we've had a chance
+                 * to set it to the user-preferred value.
+                 */
+                if (m_Current == m_UserMaximum) {
+                    m_Semaphore.notify();
+                }
+
+                // std::cout << "Max: " << m_Current << " ("
+                //           << std::get<0>(m_Limits) << ", "
+                //           << std::get<1>(m_Limits) << ")" << std::endl;
+            }
+        }
+
+        inline UnitType getCurrent()
+        {
+            m_Semaphore.waitOnce();
+            return m_Current;
+        }
+
+        inline Limits<UnitType> &getLimits()
+        {
+            m_Semaphore.waitOnce();
+            return m_Limits;
+        }
+
+    private:
+        UnitType m_Current;
+        Limits<UnitType> m_Limits;
+        Semaphore m_Semaphore;
+    };
+
+    ControllerPtr m_Device;
+    Semaphore m_Semaphore;
+    std::unique_ptr<VideoStream> m_VideoStream;
+    bool m_IsConnected = false;
+    FlightEventHandler m_FlightEventHandler = nullptr;
+    LimitValues<degree_t> m_TiltLimits;
+    LimitValues<meters_per_second_t> m_VerticalSpeedLimits;
+    LimitValues<degrees_per_second_t> m_YawSpeedLimits;
+
+    void startStreaming();
+    void stopStreaming();
+    bool onAxisEvent(const HID::JAxis axis, const float value);
+    bool onButtonEvent(const HID::JButton button, const bool pressed);
+    inline void addEventHandlers();
+    void onBatteryChanged(const ARCONTROLLER_DICTIONARY_ELEMENT_t *dict);
+    inline void createControllerDevice();
+    inline eARCONTROLLER_DEVICE_STATE getState();
+    inline eARCONTROLLER_DEVICE_STATE getStateUpdate();
+
+    // speed limits
+    inline void setMaximumTilt(const degree_t newValue);
+    inline void setMaximumVerticalSpeed(const meters_per_second_t newValue);
+    inline void setMaximumYawSpeed(const degrees_per_second_t newValue);
+
+    static void commandReceived(eARCONTROLLER_DICTIONARY_KEY key,
+                                ARCONTROLLER_DICTIONARY_ELEMENT_t *dict,
+                                void *data);
+    static int printCallback(eARSAL_PRINT_LEVEL level,
+                             const char *tag,
+                             const char *format,
+                             va_list va);
+    static void stateChanged(eARCONTROLLER_DEVICE_STATE newstate,
+                             eARCONTROLLER_ERROR err,
+                             void *data);
+}; // Bebop
+
+// We also have to give definitions for these variables, because c++ is weird
+constexpr degree_t Bebop::DefaultMaximumTilt;
+constexpr degrees_per_second_t Bebop::DefaultMaximumYawSpeed;
+constexpr meters_per_second_t Bebop::DefaultMaximumVerticalSpeed;
+
+/*
+ * Do all initialisation (including discovery) but don't actually
+ * connect to drone yet.
+ */
+Bebop::Bebop(const degrees_per_second_t maxYawSpeed,
```

Comment:
I did start making this into a libbebop, but stopped when I realised the test program wasn't compiling any faster (I guess most of the time is spent reading in all the included headers and in linking). You're right that it would be tidier though, so it's probably worth doing it just for that reason.